src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageDrivenChannelAdapter.java

```diff
@@ -271,6 +288,32 @@ public int afterShutdown() {
 		return getPhase();
 	}
 
+	protected void setAttributesIfNecessary(Object record, Message<?> message) {
+		boolean needAttributes = getErrorChannel() != null
+				&& KafkaMessageDrivenChannelAdapter.this.retryTemplate == null;
+		if (needAttributes) {
+			attributesHolder.set(ErrorMessageUtils.getAttributeAccessor(null, null));
+		}
+		if (needAttributes || KafkaMessageDrivenChannelAdapter.this.recoveryCallback != null) {
+			AttributeAccessor attributes = attributesHolder.get();
+			if (attributes != null) {
+				attributes.setAttribute(ErrorMessageUtils.INPUT_MESSAGE_CONTEXT_KEY, message);
```

Comment:
OK. There **is** `retryTemplate` but **no** `recoveryCallback`.
The `if (needAttributes || KafkaMessageDrivenChannelAdapter.this.recoveryCallback != null) ` is `false` and we don't populated `INPUT_MESSAGE_CONTEXT_KEY` and `KAFKA_RAW_DATA`.

Am I missing anything else?

On the other hand How does it work if we have `errorChannel` and `retryTemplate`.
Looks like we don't go to the `recoveryCallback` at all. See `MPS.sendMessage()`:
```
catch (RuntimeException e) {
	MessageChannel errorChannel = getErrorChannel();
	if (errorChannel != null) {
		this.messagingTemplate.send(errorChannel, this.errorMessageStrategy.buildErrorMessage(e,
				getErrorMessageAttributes(message)));
	}
	else {
		throw e;
	}
}
```